### DISCUSSION TITLE
Design with dynamic data from a CMS

### DISCUSSION THREAD

**Discussion by lexoyo** [GrapesJS/grapesjs#6052] [General]
This discussion is about the use of [the new data source API](https://github.com/GrapesJS/grapesjs/pull/6018) in a specific use case : designing websites with dynamic data coming from a CMS. I would love to have some feedback about the plugin I created (open source, open for contributions) and ideas for the integration of the new API :)

Here is [Silex Lab's DataSource plugin](https://github.com/silexlabs/grapesjs-data-source) which will benefit the new API for rendering the dynamic data on the canvas. This plugin loads GraphQL data from a data source you specify, and use the type system provided by GraphQL to let the user build expressions in the components attributes. You can test the plugin in Silex (v3.silex.me)

Here are the dynamic attributes that the plugin stores in a component:

* States (reusable "shortcuts" or variables)
* HTML attributes
* Properties (innerHtml, visibility)
* Loop
* Given the new API, I probably should add "style"

![The properties in data source plugin](https://github.com/user-attachments/assets/92a18b55-4489-4941-854b-e9533d18a39b)
 
Here is how I would fit the API in the plugin

* The plugin keeps managing all the dynamic data as components attributes
* The plugin keeps the same logic for publication of liquid templates in the HTML
* The plugin should sync the dynamic data with the API so that we can have the dynamic data rendered in the canvas

What do you think?

**Comment by lexoyo** [comment]
I gave more thoughts about this @artf @danstarns and I don't see how to render the loops

**Reply by danstarns** [reply]
@lexoyo, thank you for sharing your thoughts and ideas. We are currently in the final stages of iterating on what we have for [DataSources](https://github.com/GrapesJS/grapesjs/pull/6018). Our plan is to release it under preview, and then we'll continue iterating by adding extensions/plugins and making adjustments to support community tools like yours.

Please bear with us—features like your remote GraphQL datasource are highly valuable to the ecosystem and are definitely on our radar.

**Reply by lexoyo** [reply]
Hello @danstarns @artf 👋

I'm currently implementing live preview to the @silexlabs/data-source plugin. I need to decide if it will use the datasources api or the onRender life cycle method

Would you share your vision on that?

* why is it better to use the datasources api? The onRender route is harder than I think?
* how will loops work with the datasources api?

Have a good day

**Reply by lexoyo** [reply]
Bump @danstarns @artf (sorry :angel: )
